runtime/src/alloc.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright 2023 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const MAX_MEMORY: usize = 0x400000;
+
+#[allow(clippy::missing_safety_doc)]
+#[no_mangle]
+pub unsafe extern "C" fn sys_alloc_aligned(bytes: usize, align: usize) -> *mut u8 {
+    #[cfg(target_arch = "riscv32")]
+    extern "C" {
+        // https://lld.llvm.org/ELF/linker_script.html#sections-command
+        static _end: u8;
+    }
+
+    // Pointer to next heap address to use, or 0 if the heap has not yet been
+    // initialized.
+    static mut HEAP_POS: usize = 0;
+
+    // SAFETY: Single threaded, so nothing else can touch this while we're working.
+    let mut heap_pos = unsafe { HEAP_POS };
+
+    #[cfg(target_arch = "riscv32")]
+    if heap_pos == 0 {
+        heap_pos = unsafe { (&_end) as *const u8 as usize };
+    }
+
+    let offset = heap_pos & (align - 1);
+    if offset != 0 {
+        heap_pos += align - offset;
+    }
+
+    let ptr = heap_pos as *mut u8;
+    let (heap_pos, overflowed) = heap_pos.overflowing_add(bytes);
+
+    if overflowed || MAX_MEMORY < heap_pos {
```

Comment:
This check is sometimes more relaxed than the previous check against the stack pointer `sp`. As a result, this allocator can return a piece of memory that's in the stack.

Reply:
Yes, instead of checking against a fixed `MAX_MEMORY` constant we should check against clashing with the stack

Reply:
Can we have a fixed stack size by default?. All most all program in Linux have a maximum limit of stack size. 
It's make this code easier to implement.

Reply:
I suggest setting the default stack size, and the check in alloc is much simpler since we don't have to keep track of stack pointer, EVM default stack size 32 Kb. We can adjust our stack size to 32 Mb. What do you think? At this point, if I run a code that has > 65 Mb stack on MacOS it's crashed.

Reply:
My guess is that the bug fix doesn't really require making the stack size constant. So it's a design issue, I defer to @sjudson .

One more reason to make it constant: we need another check when we grow the stack. The next stack element might overlap with allocated memory on the heap.

Reply:
I think ideally we want the stack to be dynamically resizeable, as there are far more significant performance impacts to these parameters in a zkVM than normal (as they impact the witness size).

Reply:
I got this problem fixed. I refer to some code in embedded code. The idea is to the stack pointer (sp) and make sure the gap between `heap_pos` and `sp` is about 1024 bytes.

Reply:
I got this problem fixed. I refer to some code in embedded code. The idea is to the stack pointer (sp) and make sure the gap between heap_pos and sp is about 1024 bytes.

Reply:
> The idea is to the stack pointer (sp) and make sure the gap between heap_pos and sp is about 1024 bytes.

I'm confused seeing a constant number 1024 here. What is the reason for 1024 bytes?

Reply:
Any number will work. This let the stack and heap grow freely instead of setting them to fix numbers. 
1024 bytes is a gap number between two. So if a `slightly` overflow somewhere, it will still work instead of corrupting other memory undetected. 
Having some gaps relax with the alignment calculation for stack and heap.

Reply:
Any number will work. This let the stack and heap grow freely instead of setting them to fix numbers. 
1024 bytes is a gap number between two. So if a `slightly` overflow somewhere, it will still work instead of corrupting other memory undetected. 
Having some gaps relax with the alignment calculation for stack and heap.

Reply:
I see.

Reply:
implemented in 0944f29